File: ui/apps/web/src/lib/audio.ts
/**
 * A source's audio tracks, and which one the player opens with (#99).
 *
 * The same rules as the API's `lib/media/audio.py`, because a file the
 * browser plays itself is decided here: it opens with the track the file
 * marks, so any other needs a session.
 */

export type AudioTrack = {
    /** Among the source's audio tracks; what a switch names. */
    index: number;
    language: string | null;
    title: string | null;
    channels: number | null;
    codec: string | null;
    /** The one the source marks to open with. */
    default: boolean;
};

export function normalizeAudioTracks(raw: any): AudioTrack[] {
    if (!Array.isArray(raw)) return [];
    return raw.map((track: any) => ({
        index: track?.index ?? 0,
        // The envelope leaves out what is null.
        language: track?.language ?? null,
        title: track?.title ?? null,
        channels: track?.channels ?? null,
        codec: track?.codec ?? null,
        default: Boolean(track?.default),
    }));
}

/** The languages Settings offers, by the ISO 639-1 code a file's and a site's both fold to. */
export const AUDIO_LANGUAGES: { value: string; label: string }[] = [
    [
        "ar",
        "Arabic",
    ],
    [
        "bn",
        "Bengali",
    ],
    [
        "zh",
        "Chinese",
    ],
    [
        "cs",
        "Czech",
    ],
    [
        "da",
        "Danish",
    ],
    [
        "nl",
        "Dutch",
    ],
    [
        "en",
        "English",
    ],
    [
        "fi",
        "Finnish",
    ],
    [
        "fr",
        "French",
    ],
    [
        "de",
        "German",
    ],
    [
        "el",
        "Greek",
    ],
    [
        "he",
        "Hebrew",
    ],
    [
        "hi",
        "Hindi",
    ],
    [
        "hu",
        "Hungarian",
    ],
    [
        "id",
        "Indonesian",
    ],
    [
        "it",
        "Italian",
    ],
    [
        "ja",
        "Japanese",
    ],
    [
        "ko",
        "Korean",
    ],
    [
        "no",
        "Norwegian",
    ],
    [
        "fa",
        "Persian",
    ],
    [
        "pl",
        "Polish",
    ],
    [
        "pt",
        "Portuguese",
    ],
    [
        "ro",
        "Romanian",
    ],
    [
        "ru",
        "Russian",
    ],
    [
        "es",
        "Spanish",
    ],
    [
        "sv",
        "Swedish",
    ],
    [
        "th",
        "Thai",
    ],
    [
        "tr",
        "Turkish",
    ],
    [
        "uk",
        "Ukrainian",
    ],
    [
        "vi",
        "Vietnamese",
    ],
].map(
    ([
        value,
        label,
    ]) => ({ value, label }),
);

/** ISO 639-2 codes, bibliographic and terminological, by the ISO 639-1 code they mean. */
const TWO_LETTER: Record<string, string> = {
    ara: "ar",
    ben: "bn",
    chi: "zh",
    zho: "zh",
    cze: "cs",
    ces: "cs",
    dan: "da",
    dut: "nl",
    nld: "nl",
    eng: "en",
    fin: "fi",
    fre: "fr",
    fra: "fr",
    ger: "de",
    deu: "de",
    gre: "el",
    ell: "el",
    heb: "he",
    hin: "hi",
    hun: "hu",
    ind: "id",
    ita: "it",
    jpn: "ja",
    kor: "ko",
    may: "ms",
    msa: "ms",
    nor: "no",
    nob: "no",
    nno: "no",
    per: "fa",
    fas: "fa",
    pol: "pl",
    por: "pt",
    rum: "ro",
    ron: "ro",
    rus: "ru",
    spa: "es",
    swe: "sv",
    tam: "ta",
    tel: "te",
    tha: "th",
    tur: "tr",
    ukr: "uk",
    urd: "ur",
    vie: "vi",
    fil: "tl",
    tgl: "tl",
};

const UNKNOWN = new Set([
    "und",
    "unk",
    "mis",
    "mul",
    "zxx",
    "",
]);

/** A code's language alone, comparable across ISO 639-1, 639-2 and BCP 47; `null` when unknown. */
export function languageKey(code: string | null | undefined): string | null {
    if (!code) return null;
    const primary = code.trim().replace("_", "-").split("-")[0].toLowerCase();
    if (UNKNOWN.has(primary)) return null;
    return TWO_LETTER[primary] ?? primary;
}

/**
 * The track to open with: the preferred language's, else the one the source
 * marks, else the first. `null` when there are none.
 */
export function pickAudioTrack(
    tracks: AudioTrack[],
    language: string | null = null,
): number | null {
    if (tracks.length === 0) return null;
    const key = languageKey(language);
    if (key !== null) {
        const match = tracks.find(
            (track) => languageKey(track.language) === key,
        );
        if (match) return match.index;
    }
    return (tracks.find((track) => track.default) ?? tracks[0]).index;
}

/**
 * Whether a file the browser could play itself needs a session anyway: the
 * browser opens with the track the file marks, and the preference picks
 * another.
 */
export function needsAudioSession(
    tracks: AudioTrack[],
    language: string | null,
): boolean {
    return pickAudioTrack(tracks, language) !== pickAudioTrack(tracks, null);
}

const CHANNELS: Record<number, string> = {
    1: "mono",
    2: "stereo",
    6: "5.1",
    8: "7.1",
};

function languageName(code: string | null): string | null {
    const key = languageKey(code);
    if (key === null) return null;
    try {
        return (
            new Intl.DisplayNames(
                [
                    "en",
                ],
                { type: "language" },
            ).of(key) ?? key
        );
    } catch {
        return key;
    }
}

/**
 * How the menu names a track: its language, its title when the file gives
 * one, and its channels. Tracks that share all of that are told apart by
 * number, as is one with nothing to say.
 */
export function audioTrackLabel(
    track: AudioTrack,
    tracks: AudioTrack[] = [
        track,
    ],
): string {
    const describe = (t: AudioTrack) =>
        [
            languageName(t.language),
            t.title,
            t.channels !== null
                ? (CHANNELS[t.channels] ?? `${t.channels} ch`)
                : null,
        ]
            .filter(Boolean)
            .join(" · ");
    const label = describe(track);
    const twins = tracks.filter((other) => describe(other) === label);
    if (!label) return `Track ${track.index + 1}`;
    return twins.length > 1 ? `${label} (${track.index + 1})` : label;
}

/**
 * Which segment of an HLS media playlist holds `time`, by adding up its
 * `#EXTINF` durations. The one to have ready before swapping to a new
 * session, so the picture carries on from there. The last when `time` is
 * past the end.
 */
export function segmentAt(playlist: string, time: number): number {
    let start = 0;
    let index = -1;
    for (const line of playlist.split("\n")) {
        const match = /^#EXTINF:([\d.]+)/.exec(line.trim());
        if (!match) continue;
        index += 1;
        start += Number(match[1]);
        if (time < start) return index;
    }
    return Math.max(0, index);
}

File: api/src/lib/media/audio.py
"""A source's audio tracks, and which one a player opens with (#99).

A file names a track's language in ISO 639-2 (``eng``, and ``fre`` or ``fra``),
a site in BCP 47 (``en``, ``en-US``). They're compared by ``language_key``,
which folds the common ones onto ISO 639-1. Anything it doesn't know is
compared as it is, which still matches a code against itself.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class AudioTrack:
    #: Counted among the source's audio tracks only, as ``-map 0:a:N`` counts.
    #: For a site, its place in the page's list of audio formats.
    index: int
    language: str | None = None
    title: str | None = None
    channels: int | None = None
    codec: str | None = None
    #: The track the source marks as the one to open with.
    default: bool = False


#: ISO 639-2 codes, bibliographic and terminological, by the ISO 639-1 code
#: they mean. The languages the Settings menu offers, and a few more.
_TWO_LETTER = {
    "ara": "ar", "ben": "bn", "chi": "zh", "zho": "zh", "cze": "cs", "ces": "cs",
    "dan": "da", "dut": "nl", "nld": "nl", "eng": "en", "fin": "fi", "fre": "fr",
    "fra": "fr", "ger": "de", "deu": "de", "gre": "el", "ell": "el", "heb": "he",
    "hin": "hi", "hun": "hu", "ind": "id", "ita": "it", "jpn": "ja", "kor": "ko",
    "may": "ms", "msa": "ms", "nor": "no", "nob": "no", "nno": "no", "per": "fa",
    "fas": "fa", "pol": "pl", "por": "pt", "rum": "ro", "ron": "ro", "rus": "ru",
    "spa": "es", "swe": "sv", "tam": "ta", "tel": "te", "tha": "th", "tur": "tr",
    "ukr": "uk", "urd": "ur", "vie": "vi", "fil": "tl", "tgl": "tl",
}

#: What a file writes when it doesn't know.
_UNKNOWN = {"und", "unk", "mis", "mul", "zxx", ""}

#: The languages above by their English names, as subtitle files are often
#: named (``2_English.srt``), and a few native ones releases use.
_NAMES = {
    "arabic": "ar", "bengali": "bn", "chinese": "zh", "czech": "cs", "danish": "da",
    "dutch": "nl", "english": "en", "finnish": "fi", "french": "fr", "francais": "fr",
    "german": "de", "deutsch": "de", "greek": "el", "hebrew": "he", "hindi": "hi",
    "hungarian": "hu", "indonesian": "id", "italian": "it", "italiano": "it",
    "japanese": "ja", "korean": "ko", "malay": "ms", "norwegian": "no", "persian": "fa",
    "farsi": "fa", "polish": "pl", "portuguese": "pt", "brazilian": "pt", "romanian": "ro",
    "russian": "ru", "spanish": "es", "espanol": "es", "latino": "es", "swedish": "sv",
    "tamil": "ta", "telugu": "te", "thai": "th", "turkish": "tr", "ukrainian": "uk",
    "urdu": "ur", "vietnamese": "vi", "filipino": "tl", "tagalog": "tl",
}

#: Two-letter codes a file name can carry. Not "hi": beside a language,
#: ``Movie.en.hi.srt`` means hearing impaired far more often than Hindi.
_FILE_NAME_CODES = (set(_TWO_LETTER.values()) | {"en", "es", "pt", "it"}) - {"hi"}


def language_in_name(token: str) -> str | None:
    """The ISO 639-1 code one word of a file name stands for (``en``, ``eng``, ``English``), if any."""
    word = token.strip().lower()
    if word in _NAMES:
        return _NAMES[word]
    if word in _TWO_LETTER:
        return _TWO_LETTER[word]
    return word if word in _FILE_NAME_CODES else None


def language_key(code: str | None) -> str | None:
    """``code``'s language alone, comparable across ISO 639-1, 639-2 and BCP 47; ``None`` when unknown."""
    if not code:
        return None
    primary = code.strip().replace("_", "-").split("-", 1)[0].lower()
    if primary in _UNKNOWN:
        return None
    return _TWO_LETTER.get(primary, primary)


def pick_audio_track(
    tracks: Sequence[AudioTrack], language: str | None = None, wanted: int | None = None
) -> int | None:
    """The track to open with: ``wanted`` if there is one, else the first match below.

    1. the preferred ``language``
    2. the track the source marks as default
    3. the first track

    ``None`` when there are no tracks at all.
    """
    if not tracks:
        return None
    indexes = [track.index for track in tracks]
    if wanted is not None and wanted in indexes:
        return wanted
    key = language_key(language)
    if key is not None:
        match = next((track for track in tracks if language_key(track.language) == key), None)
        if match is not None:
            return match.index
    marked = next((track for track in tracks if track.default), None)
    return (marked or tracks[0]).index
